plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/flexible/FlexExistingArtifactDeployCommandHandler.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.appengine.deploy.ui.flexible;
+
+import com.google.cloud.tools.eclipse.appengine.deploy.DeployPreferences;
+import com.google.cloud.tools.eclipse.appengine.deploy.StagingDelegate;
+import com.google.cloud.tools.eclipse.appengine.deploy.flex.FlexExistingArtifactDeployPreferences;
+import com.google.cloud.tools.eclipse.appengine.deploy.flex.FlexExistingDeployArtifactStagingDelegate;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployCommandHandler;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.DeployPreferencesDialog;
+import com.google.cloud.tools.eclipse.appengine.deploy.ui.Messages;
+import com.google.cloud.tools.eclipse.googleapis.IGoogleApiFactory;
+import com.google.cloud.tools.eclipse.login.IGoogleLoginService;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import org.eclipse.core.commands.ExecutionEvent;
+import org.eclipse.core.commands.ExecutionException;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.swt.widgets.Shell;
+
+public class FlexExistingArtifactDeployCommandHandler extends DeployCommandHandler {
+
+  @Override
+  protected DeployPreferencesDialog newDeployPreferencesDialog(Shell shell, IProject project,
+      IGoogleLoginService loginService, IGoogleApiFactory googleApiFactory) {
+    String title = Messages.getString("deploy.preferences.dialog.title.flexible");
+    return new FlexExistingArtifactDeployPreferencesDialog(shell, title, loginService,
+        googleApiFactory);
+  }
+
+  @Override
+  protected StagingDelegate getStagingDelegate(IProject project) throws CoreException {
+    String appYamlPath = new FlexExistingArtifactDeployPreferences().getAppYamlPath();
```

Comment:
This repeated use of `new FlexExistingArtifactDeployPreferences()` seems unusual — it makes sense for the global dialog, but I think it would be good to extract a method that is passed in the artifact and app.yaml.  

I also think this handler can do double-duty, both deploying an existing artifact that's currently selected (e.g., right-clicking on a jar/war and optionally app.yaml in the _Project Explorer_), as well as to prompt for the file.

Given that you don't use the IProject argument, and from looking at DeployCommandHandler, I wonder if you might not be better off  not subclassing?

Reply:
> I also think this handler can do double-duty, both deploying an existing artifact that's currently selected (e.g., right-clicking on a jar/war and optionally app.yaml in the Project Explorer), as well as to prompt for the file.

We'll get to that point eventually (https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2103). Just not now.

Regarding the other refactoring suggestions, I did expect such comments, and I'll say this again:

> Can't say I'm 100% satisfied with the current code structure, but I decided to go for reasonably minimal code change, as otherwise refactoring starts to propagate to too many files.

To add more, I spent quite a lot of time on what would be the ideal code, and all the refactoring attempts ended up touching a lot of files and a huge PR where the majority of the diff is a distraction from what should be reviewed due to refactoring. So, if it isn't a tangible and small suggestion that can be done in this PR, I'd prefer put off big high-level restructuring ideas.

Reply:
Also for the `new FlexExistingArtifactDeployPreferences()`, we already have an issue: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2416